### PR TITLE
Update PyO3 and Maturin and require Python 3.7

### LINKED ISSUE #27: Update minimum supported Python version to 3.7
Python 3.6 is [EOL via PEP 494](https://www.python.org/dev/peps/pep-0494/#id15) as of Dec 2021.

- [ ] Change ABI feature compatibility to 3.7
- [ ] Update PyO3/Maturin
- [ ] Update documentation

### PR DESCRIPTION
**Pull request by samarsha**
Closes #27.

### CODE REVIEW

File: pyqir-parser/Cargo.toml [hunk 1]
```diff
@@ -9,21 +9,16 @@ readme = "README.md"
...
 
 [features]
-extension-module = ["pyo3/abi3-py36", "pyo3/extension-module"]
-python-bindings = []
 llvm11-0 = ["llvm-ir/llvm-11", "inkwell/llvm11-0"]
 llvm12-0 = ["llvm-ir/llvm-12", "inkwell/llvm12-0"]
 llvm13-0 = ["llvm-ir/llvm-13", "inkwell/llvm13-0"]
-#llvm14-0 = ["llvm-ir/llvm-14", "inkwell/llvm14-0"]
-default = ["extension-module", "python-bindings"]
+llvm14-0 = ["llvm-ir/llvm-14", "inkwell/llvm14-0"]
```
Comment (idavis) [inline pyqir-parser/Cargo.toml:22]: @samarsha Why is the `python-bindings`  feature removed?

File: docs/development-guide/compatibility.md [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 ## Operating systems
 
-PyQIR runs on most x86-64 operating systems that can run Python 3.6+; however,
+PyQIR runs on most x86-64 operating systems that can run Python 3.7+; however,
```
Comment (idavis) [inline docs/development-guide/compatibility.md:5]: The supported tier 1 and tier 2 need to be updated according to the system we validate as part of the build. jammy, bullseye, etc

File: pyqir-parser/Cargo.toml [hunk 3]
```diff
@@ -9,21 +9,16 @@ readme = "README.md"
...
```
Comment (idavis) [inline pyqir-parser/Cargo.toml:21]: Disable until we tackle this as a feature.

File: pyqir-parser/Cargo.toml (hunk 1 above)
Comment (samarsha) [reply pyqir-parser/Cargo.toml:24]: I added it back for pyqir-parser.

File: eng/Dockerfile.examples [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+RUN wget "https://packages.microsoft.com/config/ubuntu/22.04/packages-microsoft-prod.deb" \
+    && dpkg -i packages-microsoft-prod.deb \
+    && rm packages-microsoft-prod.deb \
+    && apt-get update \
+    && apt-get install -y --no-install-recommends powershell python3-minimal python3-pip python3-setuptools sudo \
+    && rm -rf /var/lib/apt/lists/*
+
+FROM $RELEASE
+
+ARG USERNAME=pyqir
```
Comment (idavis) [inline eng/Dockerfile.examples:43]: @samarsha Have you tried overriding these vars (`USERNAME`, `USER_ID`, `USER_GID`) with `--build-arg` to make sure they are still applied?

File: qirlib/Cargo.toml [hunk 5]
```diff
@@ -28,6 +28,10 @@ lazy_static = "1.4.0"
 mut_static = "5.0.0"
 bitvec = "1.0.0"
 
+# TODO: Avoid a transitive unicode-ident dependency for now due to licensing.
+syn = "1.0, < 1.0.95"
```
Comment (idavis) [inline qirlib/Cargo.toml:32]: @samarsha I don't understand. Does the TODO mean we still need to work around the issue, or that this is a workaround for that issue?

File: eng/Dockerfile.examples (hunk 4 above)
Comment (samarsha) [reply eng/Dockerfile.examples:43]: Yes, the build args still work.

File: qirlib/Cargo.toml (hunk 5 above)
Comment (samarsha) [reply qirlib/Cargo.toml:32]: The version upper bounds for syn and proc-macro2 is the workaround, the todo is to remove the bounds when we can allowlist the license.

Comment (samarsha) [reply qirlib/Cargo.toml:32]: Updated the comment.

File: eng/psakefile.ps1 [hunk 6]
```diff
@@ -219,7 +219,7 @@ task python-requirements -depends check-environment {
 
 task init -depends python-requirements {
     if (Test-CI) {
-        & $python -m pip install maturin==0.12.12
+        & $python -m pip install maturin~=0.13.6
```
[the remark this answers is not in the document]
Comment (samarsha) [reply eng/psakefile.ps1]: I think I would rather make the build-system.requires field in the pyproject.tomls the source of truth for Maturin. I removed the instances where we manually install/run Maturin. Pip now handles it automatically instead.

File: pyqir-evaluator/NOTICE-WHEEL.txt [hunk 7]
```diff
@@ -9103,7 +7345,7 @@ SOFTWARE.
 
 
 
-pyqir-evaluator 0.7.0a1 - SPDX: MIT - MIT License
+pyqir-evaluator 0.7.0-alpha.1 - SPDX: MIT - MIT License
```
Comment (idavis) [inline pyqir-evaluator/NOTICE-WHEEL.txt:7348]: Version changed from `pyqir-evaluator 0.7.0a1 - SPDX: MIT - MIT License` to `pyqir-evaluator 0.7.0-alpha.1 - SPDX: MIT - MIT License`. Was this intentional?

Comment (samarsha) [reply pyqir-evaluator/NOTICE-WHEEL.txt:7348]: I don't know why it changed, but the notice files are up-to-date.

**Review by idavis — APPROVED** [review]

**Pull request merged**